### PR TITLE
Add JAX backend support

### PR DESCRIPTION
**Pull request by MazenAmria**
Rewrites the JAX layer that was prototyped on origin/jax-support against the current main package layout (leap_c/{diff_mpc,parameters,torch.py,utils}), rather than reintroducing main's own restructure (#338) from that branch.

- leap_c/autograd/jax.py: wraps DiffFunction as a jax.custom_vjp callable via jax.pure_callback. Builds a fresh custom_vjp + context box per call instead of one shared at layer-construction time, so overlapping forward calls on the same layer (e.g. two rollouts before either is differentiated) can't clobber each other's backward-pass context.
- leap_c/jax.py: AcadosDiffMpcJax, the JAX counterpart to AcadosDiffMpcTorch. Matches torch's ctx-based warmstarting (forward returns/accepts a ctx) and its u0=None "free variable" semantics (a decision made in main's torch layer after the original branch forked) rather than origin's u0=zeros default, which was quietly computing Q-values instead of V-values.
- combine_differentiable_parameters_jax implemented in parameters/manager.py. Also fixes a batch_size=0 coercion bug (`batch_size or 1`) shared by the torch and jax combine methods.
- pyproject.toml: adds a jax-cuda12 extra so CUDA JAX is a declared, lockable dependency instead of a manual post-install `uv pip install` that a later plain `uv sync` would silently revert to CPU.
- Tests mirror a proportionate subset of the torch suite, plus a jax-vs-torch gradient equivalence check (wrt x0 and a differentiable parameter) that catches composition bugs unit tests on either side miss.

### CODE REVIEW

File: leap_c/autograd/jax.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    nu: int,
+) -> Callable[..., tuple[AcadosDiffMpcCtx, "jax.Array", "jax.Array", "jax.Array", "jax.Array"]]:
+    """Creates a JAX-callable function with a custom VJP wrapping a ``DiffFunction``.
+
+    The returned function accepts ``(x0, u0, p_global, p_stagewise, ctx=None)`` and returns
+    ``(ctx, sol_u0, x, u, sol_value)``, mirroring
+    :func:`leap_c.autograd.torch.create_autograd_function`: passing a ``ctx`` from a previous
+    call warmstarts the solve, and the returned ``ctx`` can be fed into the next call.
+
+    The impure acados C solver is isolated inside ``jax.pure_callback``, and gradients flow via
```
Comment (JasperHoffmann) [inline leap_c/autograd/jax.py]: maybe use not impure :)

File: leap_c/autograd/jax.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+def _to_jax(val: np.ndarray | None) -> "jax.Array | None":
+    return None if val is None else jax.numpy.asarray(val)
+
+
+def create_jax_callable(
+    fun: DiffFunction,
+    N_horizon: int,
+    nx: int,
+    nu: int,
+) -> Callable[..., tuple[AcadosDiffMpcCtx, "jax.Array", "jax.Array", "jax.Array", "jax.Array"]]:
```
Comment (JasperHoffmann) [inline leap_c/autograd/jax.py:35]: Is it required that we implement specifically for the diff mpc function. I was hoping we can just write generacially as in PyTorch :)

**Review by JasperHoffmann — CHANGES_REQUESTED** [review]
Looks good, I am wondering though wether we can simplify the jax autograd registration.

File: leap_c/autograd/jax.py (hunk 2 above)
Comment (MazenAmria) [reply leap_c/autograd/jax.py:35]: Both comments are related, `pure_callback` requires inputs/outputs of the callback to be an array with a predefined shape and dtype in order to be compiled. PyTorch doesn't do this JIT compiling step, so we do not need to provide N_horizon, nx, and nu ahead of time. Also, the context is not an array type, so the callback can't return the context object, which is the source of the impurity mentioned in the previous comment.

I've refined the docstrings to make that clearer.
